packages/runtime/src/runtime.ts

```diff
@@ -225,29 +218,35 @@ export class Runtime {
     this.ctx.gtxs = [encTx];
   }
 
-  // creates new application and returns application id
+  /**
+   * creates new application and returns application id
+   * @param flags SSCDeployment flags
+   * @param payFlags Transaction parameters
+   * @param program approval program
+   */
   async addApp (flags: SSCDeploymentFlags, payFlags: TxParams, program: string): Promise<number> {
     const sender = flags.sender;
     const senderAcc = this.assertAccountDefined(this.store.accounts.get(sender.addr));
 
     // create app with id = 0 in globalApps for teal execution
     const app = senderAcc.addApp(0, flags);
-    this.ctx.state.globalApps.set(app.id, app.attributes);
+    this.ctx.state.accounts.set(senderAcc.address, senderAcc);
+    this.ctx.state.globalApps.set(app.id, senderAcc.address);
```

Comment:
Please don't operate on `ctx` in Runtime. We are directly contradicting the note above the runtime class here. Let's be consistent with our definition: `interpreter operates on ctx (using deepCopy of store) and runtime operates on store`. 

So, we can do these updates on `runtime.store` (if needed), and then assign a deepCopy to `ctx` before `run`. eg.
```
this.store.accounts.set(senderAcc.address, senderAcc);
this.store.globalApps.set(app.id, senderAcc.address);
this.ctx.state = deepCopy(this.store) // assign a deep copy here instead of setting directly
await this.run(..)
..
```

Reply:
we cannot do that, for deployer functions(addApp, updateApp).
because what if program gets failed, you would have updated the store already.
but according to spec, we cannot do that, first ctx will be updated, then only we move to store

also this is similar to what we do in `run` first we set the ctx and then we pass it to interpreter, and if everything goes correct, we update store.(it is more clear and avoids bug)
I can add a detailed doc for more clarification.

Reply:
One other way would be moving appApp, updateApp to interpreter.

Reply:
hmm, then we need to figure something out. The current implemention seems incorrect by definition.

Reply:
> because what if program gets failed, you would have updated the store already.

Agreed. The state would be updated even if program fails. But i see in the below code, we are deleting this app from the account (since this is a zero indexed app, we only need it's attribute and set new state)

Reply:
One way would be to pass ctx to intrpreter.addApp and let it handle this and return updated ctx
Which we can update in runtime.addApp

Reply:
> > because what if program gets failed, you would have updated the store already.
> 
> Agreed. The state would be updated even if program fails. But i see in the below code, we are deleting this app from the account anyway (since this is a zero indexed app, we only need it's attribute and set new state with id appCounter + 1). Will check more

if program fails, error will be thrown therefore we won't reach that line. store will be updated.

Reply:
also i don't think we can move it to interpreter,, that will create cycle. best would be to add doc. what do you think?

Reply:
> if program fails, error will be thrown therefore we won't reach that line. store will be updated.

We can wrap the `run` in try {} catch () and throw error after deleting ?
Anyways, adding doc seems to be OK here. Please explain why are operating on `ctx` here in a doc string.

Reply:
ok